File: backend/app/repositories/recommendations.py
from __future__ import annotations

import logging
from datetime import datetime, timedelta, timezone
from typing import Any

from app.core.time import parse_iso_datetime

logger = logging.getLogger(__name__)


class RecommendationsRepository:
    def __init__(self, *, supabase: Any):
        self._supa = supabase

    @staticmethod
    def _normalize_symbol(symbol: str | None) -> str | None:
        s = str(symbol or "").strip().upper()
        return s or None

    def fetch_recommendation_rows(self, *, symbol: str | None, days: int, limit: int) -> list[dict[str, Any]]:
        """Read recommendation events from `youtuber_recommendations` joined to `videos`.

        Returns an empty list when the table is missing or on query errors.
        """

        sym = self._normalize_symbol(symbol)
        start_dt = datetime.now(timezone.utc) - timedelta(days=max(1, int(days)))

        try:
            q = (
                self._supa.table("youtuber_recommendations")
                .select(
                    "video_id,ticker,action,published_at,"
                    "videos(title,channel,published_at,video_url,thumbnail_url)"
                )
                .gte("published_at", start_dt.isoformat())
                .order("published_at", desc=True)
                .limit(max(1, min(2000, int(limit))))
            )
            if sym:
                q = q.eq("ticker", sym)
            resp = q.execute()
            rows = resp.data or []
            return [r for r in rows if isinstance(r, dict)]
        except Exception:
            # Backward compatibility: older DBs may not have youtuber_recommendations.published_at.
            try:
                q = (
                    self._supa.table("youtuber_recommendations")
                    .select(
                        "video_id,ticker,action,created_at,"
                        "videos(title,channel,published_at,video_url,thumbnail_url)"
                    )
                    .order("created_at", desc=True)
                    .limit(max(1, min(2000, int(limit))))
                )
                if sym:
                    q = q.eq("ticker", sym)
                resp = q.execute()
                rows = [r for r in (resp.data or []) if isinstance(r, dict)]

                def _get_published_dt(row: dict[str, Any]) -> datetime | None:
                    pa_row = row.get("published_at")
                    if pa_row:
                        try:
                            return parse_iso_datetime(str(pa_row)).astimezone(timezone.utc)
                        except Exception:
                            return None
                    v = row.get("videos")
                    if isinstance(v, list):
                        v = v[0] if v else None
                    if not isinstance(v, dict):
                        return None
                    pa = v.get("published_at")
                    if not pa:
                        return None
                    try:
                        return parse_iso_datetime(str(pa)).astimezone(timezone.utc)
                    except Exception:
                        return None

                start_dt_utc = start_dt.astimezone(timezone.utc)
                filtered: list[dict[str, Any]] = []
                for r in rows:
                    pub_dt = _get_published_dt(r)
                    if pub_dt is None:
                        continue
                    if pub_dt >= start_dt_utc:
                        filtered.append(r)

                filtered.sort(
                    key=lambda r: (_get_published_dt(r) or datetime.min.replace(tzinfo=timezone.utc)),
                    reverse=True,
                )
                return filtered
            except Exception as exc2:
                msg = str(exc2)
                if "youtuber_recommendations" in msg and (
                    "does not exist" in msg or "relation" in msg or "404" in msg
                ):
                    return []
                logger.exception("Failed to query youtuber_recommendations")
                return []

    def fetch_summary_rows_for_recommendations(self, *, video_ids: list[str], tickers: list[str]) -> list[dict[str, Any]]:
        video_ids_norm = [str(v).strip() for v in video_ids if str(v).strip()]
        tickers_norm = [str(t).strip().upper() for t in tickers if str(t).strip()]
        if not video_ids_norm or not tickers_norm:
            return []

        try:
            resp = (
                self._supa.table("summaries")
                .select("video_id,ticker,summary")
                .in_("video_id", video_ids_norm)
                .in_("ticker", tickers_norm)
                .limit(max(1, min(5000, len(video_ids_norm) * 4)))
                .execute()
            )
            rows = resp.data or []
            return [r for r in rows if isinstance(r, dict)]
        except Exception:
            logger.exception("Failed to query summaries for recommendations")
            return []

File: backend/internal/repo/recommendations.go
package repo

import (
	"fmt"
	"log/slog"

	"yunews/backend/internal/supabase"
)

// RecommendationsRepository accesses the youtuber_recommendations and summaries tables.
type RecommendationsRepository struct {
	db *supabase.Client
}

// NewRecommendationsRepository creates a new RecommendationsRepository.
func NewRecommendationsRepository(db *supabase.Client) *RecommendationsRepository {
	return &RecommendationsRepository{db: db}
}

// FetchRecommendationRows fetches recommendation events joined with video metadata.
// symbol may be empty to fetch all. startISO is the earliest published_at to include.
func (r *RecommendationsRepository) FetchRecommendationRows(symbol, startISO string, limit int) ([]map[string]interface{}, error) {
	// Try published_at column first; fall back to created_at for older schemas.
	rows, err := r.fetchRecsByPublishedAt(symbol, startISO, limit)
	if err == nil {
		return rows, nil
	}

	slog.Warn("recommendations: published_at query failed, trying created_at fallback", "error", err)
	rows, err2 := r.fetchRecsByCreatedAt(symbol, limit)
	if err2 != nil {
		if isTableMissingError(err2) {
			return nil, nil
		}
		slog.Error("recommendations: created_at fallback also failed", "error", err2)
		return nil, nil
	}
	return rows, nil
}

func (r *RecommendationsRepository) fetchRecsByPublishedAt(symbol, startISO string, limit int) ([]map[string]interface{}, error) {
	q := r.db.From("youtuber_recommendations").
		Select("video_id,ticker,action,published_at,videos(title,channel,published_at,video_url,thumbnail_url)").
		Gte("published_at", startISO).
		Order("published_at", true).
		Limit(limit)
	if symbol != "" {
		q = q.Eq("ticker", symbol)
	}
	var rows []map[string]interface{}
	if err := q.Execute(&rows); err != nil {
		return nil, err
	}
	return rows, nil
}

func (r *RecommendationsRepository) fetchRecsByCreatedAt(symbol string, limit int) ([]map[string]interface{}, error) {
	q := r.db.From("youtuber_recommendations").
		Select("video_id,ticker,action,created_at,videos(title,channel,published_at,video_url,thumbnail_url)").
		Order("created_at", true).
		Limit(limit)
	if symbol != "" {
		q = q.Eq("ticker", symbol)
	}
	var rows []map[string]interface{}
	if err := q.Execute(&rows); err != nil {
		return nil, err
	}
	return rows, nil
}

// FetchSummaryRowsForRecommendations fetches summaries matching video_id+ticker pairs.
func (r *RecommendationsRepository) FetchSummaryRowsForRecommendations(videoIDs, tickers []string) ([]map[string]interface{}, error) {
	if len(videoIDs) == 0 || len(tickers) == 0 {
		return nil, nil
	}

	limitN := len(videoIDs) * 4
	if limitN < 1 {
		limitN = 1
	}
	if limitN > 5000 {
		limitN = 5000
	}

	var rows []map[string]interface{}
	err := r.db.From("summaries").
		Select("video_id,ticker,summary").
		In("video_id", videoIDs).
		In("ticker", tickers).
		Limit(limitN).
		Execute(&rows)
	if err != nil {
		slog.Error("fetch summary rows for recommendations failed", "error", err)
		return nil, fmt.Errorf("fetch summary rows for recommendations: %w", err)
	}
	return rows, nil
}

func isTableMissingError(err error) bool {
	if err == nil {
		return false
	}
	msg := err.Error()
	return contains(msg, "youtuber_recommendations") && (contains(msg, "does not exist") || contains(msg, "relation") || contains(msg, "404"))
}

func contains(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr || len(substr) == 0 ||
		func() bool {
			for i := 0; i+len(substr) <= len(s); i++ {
				if s[i:i+len(substr)] == substr {
					return true
				}
			}
			return false
		}())
}
